sdks/python/apache_beam/runners/interactive/recording_manager.py

```diff
@@ -298,8 +298,12 @@ def _watch(self, pcolls):
           watched_pcollections.add(val)
         elif isinstance(val, DeferredBase):
           watched_dataframes.add(val)
-    # Convert them all in a single step for efficiency.
-    for pcoll in to_pcollection(*watched_dataframes, always_return_tuple=True):
+
+    # Convert them one-by-one to generate a unique label for each. This allows
+    # caching at a more fine-grained granularity.
+    for df in watched_dataframes:
+      pcoll = to_pcollection(
+          df, yield_elements='pandas', label=str(id(df._expr._id)))
```

Comment:
Right, I found that the default label generated from to_pcollection wasn't unique enough for notebooks. If you take a look at the InteractiveRunnerTest.test_dataframes_same_cell_twice test, then you can see that the default label is the same for both df['square'] and df['cube']. I can take a look in the _var_name method, maybe something wrong is there.

Reply:
Ah I guess the examples in `test_dataframe_same_cell_twice` are [falling through to the default case](https://github.com/apache/beam/blob/dd2f67bb8ec49e542635035743a754421cce3a75/sdks/python/apache_beam/dataframe/convert.py#L244), `ToPCollection(...)`, since the dataframes aren't assigned to variables. The label is generated for the `DataframeTransform` though, not the output PCollections. I wouldn't think it would impact PCollection caching.

Reply:
I'm getting a lot of errors when trying it out, so I think I'll investigate it after this PR.
```
Traceback (most recent call last):
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/runners/interactive/interactive_runner_test.py", line 389, in test_dataframes_with_multi_index_get_result
    pd.testing.assert_series_equal(df_expected, ib.collect(deferred_df, n=10))
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/runners/interactive/utils.py", line 247, in run_within_progress_indicator
    return func(*args, **kwargs)
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/runners/interactive/interactive_beam.py", line 579, in collect
    recording = recording_manager.record([pcoll], max_n=n, max_duration=duration)
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/runners/interactive/recording_manager.py", line 433, in record
    self._watch(pcolls)
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/runners/interactive/recording_manager.py", line 306, in _watch
    for pcoll in to_pcollection(*watched_dataframes, always_return_tuple=True):
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/dataframe/convert.py", line 196, in to_pcollection
    new_results = {p: extract_input(p)
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/transforms/ptransform.py", line 1086, in __ror__
    return self.transform.__ror__(pvalueish, self.label)
  File "/home/srohde/Workdir/beam/sdks/python/apache_beam/transforms/ptransform.py", line 587, in __ror__
    raise ValueError(
ValueError: Mixing value from different pipelines not allowed.
```

Reply:
Ack, ok that makes sense. Could you file a jira and/or a drop a TODO for this?